### PR TITLE
Replace the agent processor with the clean-room implementation

### PR DESCRIPTION
**Pull request by jonastemplestein**
The clean room from #2154 becomes THE agent processor — an in-place, flag-day replacement under slug `"agent"` (contract v5.0.0). **No historical compatibility**: the legacy event vocabulary (`llm-request-scheduled`/`started`/`completed`/`cancelled`, `loop-stopped`, the `durable-object-crashed` cancel reason) is deleted everywhere — contract, processor, UIs, replay, TUI, mobile, e2e. Old journals are not expected to parse or render. Cutover notes: `tasks/agent-processor-replacement.md`.

## The processor

The offset-identified turn loop from #2154, plus the ported production machinery in the clean-room shape: compaction (token-threshold trigger + summary turn through the injected transport, `replacesHistoryThrough` barrier fold), `summary-updated` fold with the conditional waiting-clear race rule, `web-message-sent` mirroring into context, `token-usage-reported` emission with the settled batch + lifetime totals, prompt building (protocol prompt, system-prompt gate, trust demotion across the full actor taxonomy — slack/telegram/email/github/integration demote to user; agent/script stay developer — timestamp last, file preparation via `resolveModelFileUrl`), the Workers AI transport behind the injected seam (unified + BYOK gateway, raw chunk objects, usage normalization), script extraction with multi-block/malformed feedback and spill-to-workspace, and `AgentLiveState` runtime derivation (no scheduled/started phases — the one open request counts as `requested`).

## UIs render only the new vocabulary

The shared `agent-ui-reducer` (web feed, TUI, mobile, browser projector all inherit) has `llm-request-settled` as its ONLY terminal handler; `"expired"` is a first-class cancel reason with its own step label; the crash-recovery activity machinery (retry counts, restart-pending states, `durable-object-crashed`) is deleted. The web stop button appends an interrupt-policy context item instead of the dead cancelled event — cancellation is a property of new input. The replay panel reads outcomes only from the settled fact (time-to-first-chunk anchors on the requested event; the model has no started event).

**Processor revivals render as a subtle marker**: a crash mid-turn cancels nothing (the open request is adopted and settles normally), so the feed shows the `stream/processor-revived` fact itself — a hairline one-line row with a tooltip explaining the work was adopted and continued. Web, TUI, and mobile.

## Edge-case coverage carried over

The deleted legacy suites (`agent-processors.test.ts`, 3156 lines; `agent-eviction-recovery.test.ts`, 647 lines) were audited scenario-by-scenario against the step-harness suite; every still-applicable edge case (27) was ported as step-harness tests — script feedback/spill edges, fence-inside-string extraction, queue-behind-open-request, interrupt-during-debounce, backoff ladder arithmetic with exact virtual-time boundaries, trust demotion taxonomy, compaction fold edges (malformed cutoff, mid-barrier arrivals, keyed system collapse, zero-dial refold of a settled journal), token tally components, and the settlement dedupe races. The processor suite is 64 step-harness tests; full apps/os is 216 files / 2045 tests.

Validation: typecheck, lint, knip, format clean; itx-api regenerated; repo-wide grep for the dead event types returns nothing.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Core agent turn loop, journal event vocabulary, and all UIs/e2e that consume agent streams change in one cutover with no backward compatibility for old journals.
> 
> **Overview**
> **Flag-day replacement** of the agent processor under contract **v5.0.0**: the legacy LLM lifecycle (`llm-request-scheduled` / `started` / `completed` / `cancelled`, `loop-stopped`, `durable-object-crashed`) is removed end-to-end. A turn is **`llm-request-requested`** (identity = journal offset) and closes with a single **`llm-request-settled`** (`succeeded` | `failed` | `cancelled`, including **`expired`**).
> 
> **Fold & runtime shape** moves to **`contextItems`**, **`openRequest`**, **`pendingLlmRequestTrigger`**, and **`agent/paused` / `resumed`**; presence/runtime no longer models scheduled/started phases. Default system prompt and config defaults shift into **`agent-defaults`** / contract defaults; the v5 **processor pretty-state** renderer keys off `contextItems` and `openRequest`.
> 
> **Feeds (web, mobile, shared reducer)** treat settlement as the only terminal LLM fact, drop crash-restart activity UX (retries, “agent restarted”), and add a **`processor-revived`** hairline when `stream/processor-revived` fires—work is adopted, not cancelled. **E2e helpers/tests** seed and assert **`llm-request-settled`** with **`requestOffset`** and updated processor snapshots.
> 
> Large legacy processor/eviction suites are deleted; behavior is covered by updated reducer/feed tests and the step-harness suite described in the PR notes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0e729af8c3cf402bc840d1a4b8de807dd8db0b54. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +2,298 -3,436 | +1,698 -2,505 🟩🟥🟥⬜⬜ |
| Mobile | +28 -24 | +25 -21 🟩⬜⬜⬜⬜ |
| UI components | +192 -144 | +162 -140 🟩⬜⬜⬜⬜ |
| Tests | +3,100 -4,832 | +2,583 -4,160 🟩🟩🟥🟥🟥 |
| Docs | +95 -0 | +87 -0 🟩⬜⬜⬜⬜ |
| Other | +0 -6 | +0 -1 🟥⬜⬜⬜⬜ |
| Generated | +384 -918 | +376 -910 🟥⬜⬜⬜⬜ |
| **Total** | **+6,097 -9,360** | **+4,931 -7,737** 🟩🟩🟥🟥🟥 |

<sub>Lines counts every changed line; Significant ignores blank lines and JS comments. Between 6667101 and 0e729af, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `0e729af` | [https://auth.iterate-preview-2.com](https://auth.iterate-preview-2.com) | 808.0 KiB | 19.9s | 14.7s |  | 1.8s | [Workflow run](https://github.com/iterate/iterate/actions/runs/587486699768613) | 2026-07-20T22:25:01.224Z | Preview app released. |
| Dummy Petshop | released | `0e729af` | [https://dummy-petshop.iterate-preview-2.com](https://dummy-petshop.iterate-preview-2.com) | 198.2 KiB | 7.8s | 5.5s |  | 463ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/587486699768613) | 2026-07-20T22:24:59.907Z | Preview app released. |
| OS | released | `0e729af` | [https://os.iterate-preview-2.com](https://os.iterate-preview-2.com) | 4.52 MiB (+1.1 KiB vs main) | 110.3s | 448.2s | 4 retried: is MITM-intercepted and routed through project egress with secret substitution (vitest x1) · concurrent long-running itx scripts all complete (vitest x1) · ephemeral events are raw-readable and delivered live, but never replayed to a later subscription (vitest x1) · runs "journal-is-the-record" through the project REPL (specs x1) | 17.6s | [Workflow run](https://github.com/iterate/iterate/actions/runs/587486699768613) | 2026-07-20T22:24:58.418Z | Preview app released. |
| Semaphore | released | `0e729af` | [https://semaphore.iterate-preview-2.com](https://semaphore.iterate-preview-2.com) | 441.8 KiB | 13.2s | 20.2s |  | 901ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/587486699768613) | 2026-07-20T22:24:41.764Z | Preview app released. |
| Streams Example App | released | `0e729af` | [https://streams.iterate-preview-2.com](https://streams.iterate-preview-2.com) | 1.48 MiB | 13.6s | 57.2s |  | 8.9s | [Workflow run](https://github.com/iterate/iterate/actions/runs/587486699768613) | 2026-07-20T22:24:49.803Z | Preview app released. |

</details>

### CODE REVIEW

File: packages/ui/src/components/events/agent-ui-reducer.ts [hunk 1]
```diff
@@ -761,6 +764,41 @@ function reduceAgentUiEvent(
...
+                  ? { durationMs: Math.max(0, timestampMs - step.startedAtMs) }
+                  : {}),
+              ...(usage.input == null ? {} : { inputTokens: usage.input }),
+              ...(usage.output == null ? {} : { outputTokens: usage.output }),
+              ...(errorMessage == null ? {} : { errorMessage }),
+              ...(cancelReason == null ? {} : { cancelReason }),
+            },
+      );
+    }
+
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply packages/ui/src/components/events/agent-ui-reducer.ts:736]: Real gap — fixed: the reducer now renders `agent/paused` and `agent/resumed` (the turn-loop breaker's facts) as the same pause/resume marker rows as the stream-level pair, so they appear in the web feed, TUI, mobile, and browser projector. Test added covering both events with their reasons. (811306725)

File: apps/os/src/domains/agents/agent-processor-contract.ts [hunk 2]
```diff
@@ -943,56 +676,67 @@ export const AgentProcessorContract = defineProcessorContract({
...
-    // settle expired intents. Reduce ignores it, and it is absent from
-    // `emits`: the recovery adapter appends it raw, as the runtime speaking.
-    STREAM_PROCESSOR_REVIVED_EVENT_TYPE,
+    // Every error on the stream — the processor's own emissions, the runner's
+    // poison skips, anything else — is transcribed into model-visible context.
+    "events.iterate.com/stream/error-occurred",
+    // The platform revival fact: its ordinary delivery at head is the
+    // guaranteed turn where processEvent finds and re-runs orphaned work after
+    // an eviction.
+    "events.iterate.com/stream/processor-revived",
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/agents/agent-processor-contract.ts:727]: Not a bug — the requirement this cites was explicitly revoked by Jonas in review: there is NO historical-journal compatibility requirement for this cutover. Old journals are not expected to parse, refold, or render; the legacy bridge was deliberately deleted in the same round. The task doc's locked-decisions section (which still stated the old refold landmines and misled this finding) is updated to record the ruling. (811306725)

File: packages/ui/src/components/events/agent-ui-reducer.ts [hunk 3]
```diff
@@ -540,11 +526,9 @@ function emitItem(state: AgentUiState, items: AgentUiItem[], item: AgentUiItem):
 // ---------------------------------------------------------------------------
 
 const AGENT_LLM_REQUEST_REQUESTED = "events.iterate.com/agent/llm-request-requested";
-const AGENT_LLM_REQUEST_COMPLETED = "events.iterate.com/agent/llm-request-completed";
-const AGENT_LLM_REQUEST_CANCELLED = "events.iterate.com/agent/llm-request-cancelled";
+const AGENT_LLM_REQUEST_SETTLED = "events.iterate.com/agent/llm-request-settled";
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply packages/ui/src/components/events/agent-ui-reducer.ts:529]: Not a bug — same revoked premise as the sibling thread: "old journals render forever" was the earlier plan, and Jonas explicitly overruled it (no historical compatibility; UIs render only the new vocabulary). The legacy reducer arms were deleted deliberately, and every edge case those paths' tests proved was ported to the new vocabulary. Task doc updated to record the ruling. (811306725)

File: apps/os/e2e/vitest/itx-test-support.ts [hunk 4]
```diff
@@ -41,27 +42,22 @@ export async function appendSyntheticProviderOutput(
...
-        asserted(llmRequestOffset + 3, {
-          type: "events.iterate.com/agent/llm-request-completed",
-          idempotencyKey: `agent/llm-request-completed@${llmRequestOffset}`,
+        asserted(llmRequestOffset + 2, {
+          type: "events.iterate.com/agent/llm-request-settled",
+          idempotencyKey: `agent/llm-request-settled@${llmRequestOffset}`,
           payload: {
+            requestOffset: llmRequestOffset,
             durationMs: 0,
-            llmRequestOffset,
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/e2e/vitest/itx-test-support.ts:63]: Real — the helper predates the fold-guard: prod deliberately started any raw-appended requested event, but the new fold only opens a request when a trigger is pending, so the whole synthetic turn was inert. Fixed in the helper itself: the atomic batch now LEADS with a trigger-bearing developer item (no actor → external trigger); the requested event consumes it in the same frame, and the settled fact closes the request before the at-head pass — so the processor neither dials a real provider nor journals a stray debounced intent. A step-harness test now pins the exact batch contract (fold, script extraction firing, zero dials, no stray intent after every window clears), so the e2e helper cannot silently rot again. (e22a65050)

File: packages/iterate/src/stream-tui/feed-format.ts [hunk 5]
```diff
@@ -19,13 +18,11 @@ export function formatStepLine(step: AgentUiStep): string {
...
-        : step.cancelReason === "interrupted-by-user-input"
-          ? "Stopped for your new message"
-          : step.outcome === "cancelled"
-            ? "Request cancelled"
-            : (step.model ?? "LLM request");
+      : step.cancelReason === "interrupted-by-user-input"
+        ? "Stopped for your new message"
+        : step.outcome === "cancelled"
+          ? "Request cancelled"
+          : (step.model ?? "LLM request");
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply packages/iterate/src/stream-tui/feed-format.ts]: Fixed — the TUI step line now reads "Request expired" and the mobile footer says "expired" (both keeping the model/token/duration parts), matching the web feed's first-class treatment; TUI label test added. (pushed with this round)

File: packages/ui/src/components/events/agent-ui-reducer.ts [hunk 6]
```diff
@@ -716,47 +693,36 @@ function reduceAgentUiEvent(
...
       const payload = readPayloadRecord(event);
-      const result = isRecord(payload?.result) ? payload.result : undefined;
-      const status = typeof result?.status === "string" ? result.status : "success";
+      const requestOffset = payload?.requestOffset;
+      if (payload == null || typeof requestOffset !== "number") return state;
+      const result = isRecord(payload.result) ? payload.result : undefined;
+      const status = typeof result?.status === "string" ? result.status : "succeeded";
       const usage = readUsageTokens(result?.usage);
       const errorMessage =
-        isRecord(result?.error) && typeof result.error.message === "string"
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply packages/ui/src/components/events/agent-ui-reducer.ts:727]: Real — the settled fact's `partialText` is indeed the ONLY durable record of the streamed partial (chunks are ephemeral, and the interrupt's assistant item deliberately omits `llmRequestOffset` so script extraction can never run on a half response). Fixed in both durable readers: the shared reducer now fills the step's `responseText` from `partialText` when no streamed text survived the rebuild (so `interruptedWithPartialResponse` is true again after a refresh / TUI / mobile refold), and the replay panel falls back to it for the response body. Tests: a reducer rebuild-without-chunks scenario and a replay fallback case. (aadf23d66)

File: apps/os/src/routes/_app/projects/$projectSlug/agents/streams/$.tsx [hunk 7]
```diff
@@ -135,14 +135,19 @@ function ProjectAgentDetailContent() {
...
     await itx.streams.get(streamPath).append({
-      type: "events.iterate.com/agent/llm-request-cancelled",
+      type: "events.iterate.com/agents/context-added",
       payload: {
-        phase: "requested",
-        llmRequestOffset,
-        reason: "interrupted-by-user-input",
+        role: "developer",
+        content: "The user interrupted the in-progress response from the web chat.",
+        llmRequestPolicy: { behaviour: "interrupt-current-request" },
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/routes/_app/projects/$projectSlug/agents/streams/$.tsx:154]: Real — a no-actor developer item counts as the agent loop's own feedback in `contextTriggerSource`, so the Stop didn't refill the autonomous budget and could collide with the breaker exactly as described. The Stop item now carries `actor: { type: "user", origin: "web" }`: an external trigger (budget refills, breaker can never trip on it) that still stays out of the chat feed (the reducer only renders developer items from integration actors) and is demoted to user role at prompt time. A step-harness test pins the precise race — a breaker-tripping self-trigger and the user's stop landing in ONE frame: budget resets, the stop's own turn runs, no pause. (aadf23d66)

File: apps/os/src/domains/agents/agent-processor-implementation.ts [hunk 8]
```diff
@@ -121,332 +150,618 @@ export class AgentProcessor extends StreamProcessor<AgentProcessorContract, Agen
...
+        runInBackground(() =>
+          append({
+            type: "events.iterate.com/agent/resumed",
+            payload: { reason: "external input" },
+            idempotencyKey: this.idempotencyKey(`resume/${trigger.offset}`),
+          }),
+        );
+      }
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/agents/agent-processor-implementation.ts:470]: Real — fixed. The breaker itself only ever pauses when nothing is open, but agent/paused is operator/script-appendable while a request IS open, and the paused branch returned before the adopt/expire branch, so a revived incarnation could neither run nor expire it until external input resumed. Pause now suppresses only NEW scheduling; an already-open request is a committed obligation the adopt/expire branch still settles — which matches a live incarnation, where a background attempt running when the pause lands drains to completion. Two regression tests (adopt-across-eviction-while-paused, and expire-past-horizon-while-paused), both verified to FAIL without the fix. (0e729af8c)

File: apps/os/src/domains/agents/agent-processor-contract.ts [hunk 9]
```diff
@@ -1,627 +1,330 @@
...
+      .default([])
+      .meta({
+        description:
+          "The conversation fold: ONE ordered list of every model-visible item (system items " +
+          "sit in place — providers accept system/developer content mid-history). A keyed item " +
+          "no request has covered yet is replaced in place by an update with the same key; a " +
+          "covered one appends a new occurrence (see lastLlmRequestOffset). Compaction " +
+          "restructures the list: system items move to the front (latest occurrence per key), " +
+          "the summary follows, then everything after the barrier.",
+      }),
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/agents/agent-processor-contract.ts:180]: Real — the Pretty renderer keyed off the v4 fold (`context`/`currentRequest`/`publishedThrough`/`requestGeneration`), so `asAgentState` never matched v5 and every agent fell back to raw JSON. Rewritten to the v5 shape: recognition on `contextItems`+`config`, `openRequest` (single slot, no phase enum), `lastLlmRequestOffset` for the published stat, a first-class Paused panel, and `contextItems` split by role for the history/system sections (items are now `{ offset, payload }`). A render test (built from the contract's own default fold, so it tracks the real shape) asserts the pretty stats appear for v5 state and a non-agent state still falls back. (0e729af8c)

File: apps/os/e2e/vitest/itx-test-support.ts [hunk 10]
```diff
@@ -39,29 +40,39 @@ export async function appendSyntheticProviderOutput(
...
+    const llmRequestOffset = triggerOffset + 1;
     try {
       const [, , assistantContext] = await stream.append(
+        asserted(triggerOffset, {
+          type: AGENT_CONTEXT_ADDED_TYPE,
+          payload: {
+            role: "developer",
+            content: "[e2e synthetic provider turn: the next assistant output is injected]",
+            llmRequestPolicy: { behaviour: "after-current-request" },
+          },
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/e2e/vitest/itx-test-support.ts:62]: Real, and my comment was the lie — a no-actor developer item is agent-loop, not external. The helper's lead-in now carries `actor: { type: "user", origin: "web" }`, a genuine external trigger that refills the autonomous budget instead of burning it every synthetic turn; comment corrected, and the mirroring unit test updated to match. (0e729af8c)

File: apps/os/src/domains/agents/agent-processor-implementation.ts [hunk 11]
```diff
@@ -121,332 +150,625 @@ export class AgentProcessor extends StreamProcessor<AgentProcessorContract, Agen
...
+                  payload: {
+                    llmRequestOffset: requestOffset,
+                    model: open.model,
+                    maxContextTokens: contextWindowTokens(open.model),
+                    ...usage,
+                  },
+                  idempotencyKey: this.idempotencyKey(`token-usage@${requestOffset}`),
+                },
+              ] satisfies EmittedInput<AgentProcessorContract>[])),
+        ]);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/agents/agent-processor-implementation.ts:668]: Real — fixed in follow-up #2165: the resolved completion is recorded as the in-flight partial before the success append is awaited, so an interrupt racing that append settles cancelled with the full text even when the transport never streamed a chunk. Regression test constructs exactly this window (success batch parked at the substrate, interrupt wins the settle key) and is bite-verified against the pre-fix code.
